d3log/src/batch.rs

```diff
@@ -0,0 +1,173 @@
+use differential_datalog::{DeltaMap, program::RelId, program::Update, ddval::DDValue};
+use mm_ddlog::*;
+use serde::{de, de::SeqAccess, de::Visitor, ser::SerializeTuple, Serialize, Deserialize, Serializer, Deserializer};
+use std::convert::TryFrom;
+use std::fmt;
+use std::fmt::Display;
+use std::io::{Error,ErrorKind};
+use std::collections::BTreeMap;
+
+// the other choice here would be a Vec<Update>?
+// need multiplicity
+// #[derive(Serialize, Deserialize)] - deltamap isn't serialize
+pub struct Batch {
+    // one might consider augmenting the values w/ t - kind of a
+    // performance problem for multisets
+    timestamp: u64, 
+    pub b: DeltaMap<differential_datalog::ddval::DDValue>,
+    // timestamp
+}
+
+impl Serialize for Batch {
+    // i would _like_ to expose an interface that used the names for relations
```

Comment:
the issue isn't being to map the names, its that the compiler generates the deserialize path for UpdateSerializer, and that one demultiplexes on the ids. we'd have to generate another version or think of something else

Reply:
_oh_, right, this is actually important. based on current discussion, @ is really a kind of generic IPC mechanism. since we are explicitly allowing communication between different ddlog domains (compilations), it can't be the case that the compiler assigned enumeration is sufficient to locate the target relation. this is fundamentally a dynamic linkage, we should also discuss whether relation names are really sufficient (do we need to include schema, versioning, .. ?)

Reply:
Right.  This was originally designed to make a single DDlog program known at compile time run in a distributed fashion.  In order to dynamically instantiate and connect different programs we'd need more infrastructure.  Relation names/ids is just part of the problem. We'd also need to somehow enforce type safety without relying on the compiler.  Not in this PR I guess... Let's brainstorm this separately.